include/ppx/metrics.h

```diff
@@ -0,0 +1,198 @@
+
+enum class MetricInterpretation
+{
+    NONE,
+    HIGHER_IS_BETTER,
+    LOWER_IS_BETTER,
+};
+
+struct Range
+{
+    double lowerBound = std::numeric_limits<double>::min();
+    double upperBound = std::numeric_limits<double>::max();
+};
+
+struct MetricMetadata
+{
+    std::string          name;
+    std::string          unit;
+    MetricInterpretation interpretation = MetricInterpretation::NONE;
+    Range                expectedRange;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+enum class MetricType
+{
+    GAUGE,
+    COUNTER,
+};
+
+class Metric
+{
+    friend class Run;
+
+public:
+    const MetricMetadata& GetMetadata() const;
+    MetricType            GetType() const;
+
+protected:
+    Metric(const MetricMetadata& metadata, MetricType type);
+    virtual ~Metric();
+
+private:
+    MetricMetadata mMetadata;
+    MetricType     mType;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+struct GaugeStatistics
+{
+    double min;
+    double max;
+    double average;
+    double median;
+    double standardDeviation;
+    double timeRatio;
+    double percentile90;
+    double percentile95;
+    double percentile99;
+};
+
+class MetricGauge final : public Metric
+{
+public:
+    MetricGauge(const MetricMetadata& metadata);
+
+    void   RecordEntry(double seconds, double value);
+    size_t GetEntriesCount() const;
+    void   GetEntry(size_t index, double& seconds, double& value) const;
+
+    const GaugeStatistics GetStatistics(bool realtime) const;
+
+private:
+    struct TimeSeriesEntry
+    {
+        double seconds;
+        double value;
+    };
+
+private:
+    ~MetricGauge();
+    METRICS_NO_COPY(MetricGauge)
+
+    void UpdateRealTimeStatistics(double seconds, double value);
+
+private:
+    std::vector<TimeSeriesEntry> mTimeSeries;
+    GaugeStatistics              mRealTimeStatistics;
+    double                       mAccumlatedValue;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+class MetricCounter final : public Metric
+{
+    friend class Run;
+
+public:
+    MetricCounter(const MetricMetadata& metadata);
+    uint64_t Increment(uint64_t add);
+    uint64_t Get() const;
+
+private:
+    ~MetricCounter();
+    METRICS_NO_COPY(MetricCounter)
+
+private:
+    uint64_t mCounter;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+class Run final
+{
+    friend class Manager;
+
+public:
+    Result       AddMetricGauge(MetricGauge*& outMetric, MetricMetadata metadata);
+    MetricGauge* GetMetricGauge(const char* name) const;
+
+    Result         AddMetricCounter(MetricCounter*& outMetric, MetricMetadata metadata);
+    MetricCounter* GetMetricCounter(const char* name) const;
+
+private:
+    Run(const char* name);
+    ~Run();
+    METRICS_NO_COPY(Run)
+
+    Metric* GetMetric(const char* name) const;
```

Comment:
related to my comment around the `Get` function existence. Going further, shouldn't the name simply be part of the metadata, which seem to be centered around human interpretation of the metric?
The only handle being the pointer returned by AddMetric?

Reply:
GetMetric is gone after the refactoring of the API.